sdk/azcore/policy_retry.go

```diff
@@ -55,14 +60,31 @@ var (
 	}
 )
 
-// DefaultRetryOptions returns an instance of RetryOptions initialized with default values.
-func DefaultRetryOptions() RetryOptions {
-	return RetryOptions{
-		StatusCodes:   StatusCodesForRetry,
-		MaxRetries:    defaultMaxRetries,
-		TryTimeout:    1 * time.Minute,
-		RetryDelay:    4 * time.Second,
-		MaxRetryDelay: 120 * time.Second,
+// init sets any default values
+func (o *RetryOptions) init() {
+	if o.MaxRetries == 0 {
+		o.MaxRetries = defaultMaxRetries
+	} else if o.MaxRetries < 0 {
+		o.MaxRetries = 0
+	}
+	if o.MaxRetryDelay == 0 {
+		o.MaxRetryDelay = 120 * time.Second
+	} else if o.MaxRetryDelay < 0 {
+		// not really an unlimited cap, but sufficiently large enough to be considered as such
+		o.MaxRetryDelay = math.MaxInt64
+	}
+	if o.RetryDelay == 0 {
+		o.RetryDelay = 4 * time.Second
+	} else if o.RetryDelay < 0 {
```

Comment:
I'm OK with 0s being set to our desired values but I'm not on board with the error checking and other value changing.
I'm afraid, that we silently change a value and now unexpected behavior occurs and the customer has no way to figure it out unless they step through our code. I think bad values should just remain bad values; most likely something will fail, the customer will see the failure, see the value, and correct their code.

This is also hard to maintain, may be inconsistently applied, and may hurt versioning if we ever introduce some sentinel value to have some special meaning for some field.

Reply:
This behavior is explicitly documented with this change, so I disagree that we're silently changing it.

> most likely something will fail, the customer will see the failure, see the value, and correct their code.

In trivial apps I can buy this.  However in larger, more complex apps it might not be that straight-forward.

Reply:
The only place where we _do_ silently change the value is `TryTimeout` which I can remove (zero or negative values here yield the same result, a deadline exceeded error).

Reply:
> This is also hard to maintain, may be inconsistently applied, and may hurt versioning if we ever introduce some sentinel value to have some special meaning for some field.

You do have a point here, and I contend that using the variadic approach alleviate these concerns since the values are changed _only_ when a customer explicitly calls a `WithOption()` method.

Reply:
What if a variadic function is called and passed bad values? Would you return an error or silently correct? Neither of these options is good in my opinion.

Reply:
We'd need to keep the same semantics as in this change, i.e. zero for default and less than zero for none.